File: 백준/Gold/10711. 모래성/모래성.py
import sys
from collections import deque

H, W = map(int, sys.stdin.readline().split())
input_int = lambda x: int(x) if x.isdigit() else 0
board = [list(map(input_int, list(sys.stdin.readline()))) for _ in range(H)]


def pb(board):
    for i in range(len(board)):
        print(board[i])
    print("")


# pb(board)

que = deque()
dr = [-1, -1, -1, 0, 0, 1, 1, 1]
dc = [-1, 0, 1, -1, 1, -1, 0, 1]
answer = 0

for i in range(H):
    for j in range(W):
        if board[i][j] == 0:
            que.append((i, j, 0))


def oob(r, c):
    return not (0 <= r < H and 0 <= c < W)


while que:
    r, c, cnt = que.popleft();
    answer = max(answer, cnt)

    for i in range(8):
        nr = r + dr[i]
        nc = c + dc[i]
        if oob(nr, nc): continue
        if board[nr][nc] == 0: continue
        board[nr][nc] -= 1
        if board[nr][nc] == 0:
            que.append((nr, nc, cnt + 1))

print(answer)

File: 백준/Gold/10711. 모래성/모래성.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.Arrays;
import java.util.LinkedList;
import java.util.Queue;
import java.util.StringTokenizer;

public class Main {
    private static BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
    private static int H, W;
    private static int[][] board;
    private static final int[] dr = new int[] {-1, -1, -1, 0, 0, 1, 1, 1};
    private static final int[] dc = new int[] {-1, 0, 1, -1, 1, -1, 0, 1};
    private static final int DIRECTION_COUNT = 8;

    public static void main(String[] args) throws IOException {
        parseInput();
//        printBoard();
        solution();
    }

    private static void printBoard() {
        for (int i = 0; i < H; i++) {
            System.out.println(Arrays.toString(board[i]));
        }
    }

    private static void parseInput() throws IOException {
        StringTokenizer st = new StringTokenizer(br.readLine());
        H = Integer.parseInt(st.nextToken());
        W = Integer.parseInt(st.nextToken());
        board = new int[H][W];
        for (int i = 0; i < H; i++) {
            String input = br.readLine();
            for (int j = 0; j < W; j++) {
                if (input.charAt(j) == '.') {
                    board[i][j] = 0;
                } else {
                    board[i][j] = Character.getNumericValue(input.charAt(j));
                }
            }
        }
    }

    private static void solution() {
        int answer = 0;
        Queue<State> que = new LinkedList<>();
        for (int i = 0; i < H; i++) {
            for (int j = 0; j < W; j++) {
                if (board[i][j] == 0) {
                    que.add(new State(i, j, 0));
                }
            }
        }

        while (!que.isEmpty()) {
            State cur = que.poll();
            answer = Math.max(answer, cur.s);
            for (int i = 0; i < DIRECTION_COUNT; i++) {
                int nr = cur.r + dr[i];
                int nc = cur.c + dc[i];
                if (oob(nr, nc)) continue;
                if (board[nr][nc] == 0) continue;
                if (--board[nr][nc] == 0) {
                    que.add(new State(nr, nc, cur.s + 1));
                }
            }
        }
        System.out.println(answer);
    }

    private static boolean oob(int r, int c) {
        return r < 0 || H <= r || c < 0 || W <= c;
    }

    static class State {
        public int r, c, s;

        public State (int r, int c, int s) {
            this.r = r;
            this.c = c;
            this.s = s;
        }
    }
}
